### ISSUE TITLE
Where is the part where the mel spectrogram is used for training?

### ISSUE THREAD

**Issue by clairerity** [ming024/FastSpeech2#141]
Hello! Thanks for this wonderful repo. I would just like to ask where the mel spectrogram files from the dataset will be input and used for training? I see that pitch, duration and energy are used in VarianceAdaptor as pitch_target, duration_target and energy_target respectively and the phonemes are used as input for the encoder, but i don't see where the mel spectrogram comes in? I don't see a mel_target or anything related to it. Thanks in advance!

**Comment by chenming6615** [comment]
Hi! The ground truth of mel spectrogram data is used here as the  `mel_targets` variable.

``` python
    def forward(self, inputs, predictions):
        (
            mel_targets,
            _,
            _,
            pitch_targets,
            energy_targets,
            duration_targets,
        ) = inputs[6:]
```

https://github.com/ming024/FastSpeech2/blob/d4e79eb52e8b01d24703b2dfc0385544092958f3/model/loss.py?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L21

It is used to  compute the loss with the predicted mel  spectrogram.

``` python
mel_loss = self.mae_loss(mel_predictions, mel_targets)
```

**Comment by clairerity** [comment]
Thank you very much for your reply @chenming6615 !! Appreciate this a lot. I just have one last question, what i don't understand is that when you load your data from the dataloader, the order of the data is this:

https://github.com/ming024/FastSpeech2/blob/d4e79eb52e8b01d24703b2dfc0385544092958f3/dataset.py#L112-L125

and when you try to input it to the model, the order becomes like this:

https://github.com/ming024/FastSpeech2/blob/d4e79eb52e8b01d24703b2dfc0385544092958f3/model/fastspeech2.py#L43-L58

what i mean is ex. how would the model know that "pitches" from the dataloader corresponds to "p_targets" or "text_lens" corresponds to "src_lens" when they were never declared as that? The order is also different, for ex. "speakers" in dataloader is the 3rd element of the list whereas in the model it is the first input. Thank you very much in advance! Have a nice day ahead :)

**Comment by chenming6615** [comment]
https://github.com/ming024/FastSpeech2/blob/d4e79eb52e8b01d24703b2dfc0385544092958f3/train.py#L82

output = model(*(batch[2:]))

Because the output of dataloader is sliced from position 2 before it is inputted into the model's forward.
So the 3rd element of the output of dataloader becomes the 1st in the model's input.

see https://www.geeksforgeeks.org/python-list-slicing/

By the way, I am not the author of this project. I just read and debugged some of the code. :)

**Comment by clairerity** [comment]
hello @chenming6615 ! thank you very much for this!! this is exactly the answer to my question. thank you once again :D I'll be closing this issue already :)

**Comment by leslie2046** [comment]
mark
